src/mqt/bench/benchmark_generator.py

```diff
@@ -3,429 +3,382 @@
-        "vqe": "mqt.bench.benchmarks.vqe",
-        "wstate": "mqt.bench.benchmarks.wstate",
-    }
-
-
-def create_benchmarks_from_config(cfg_path: str = "./config.json") -> bool:
-    init_module_paths()
-
-    with Path(cfg_path).open() as jsonfile:
-        cfg = json.load(jsonfile)
-        print("Read config successful")
-
-    global timeout
-    timeout = cfg["timeout"]
-
-    Path(utils.get_qasm_output_path()).mkdir(exist_ok=True, parents=True)
-
-    Parallel(n_jobs=-1, verbose=100)(delayed(generate_benchmark)(benchmark) for benchmark in cfg["benchmarks"])
-    return True
-
-
-def benchmark_generation_watcher(func, args):
-    class TimeoutException(Exception):  # Custom exception class
-        pass
+from mqt.bench.benchmarks import (
+    groundstate,
+    hhl,
+    pricingcall,
+    pricingput,
+    routing,
+    shor,
+    tsp,
+)
+
+
+class BenchmarkGenerator:
+    def __init__(self, cfg_path: str = "./config.json", qasm_output_path: str | None = None):
+        with Path(cfg_path).open() as jsonfile:
+            self.cfg = json.load(jsonfile)
+            print("Read config successful")
+        self.timeout = self.cfg["timeout"]
+        self.error_msg_outer_loop = "\n Problem occurred in outer loop: "
+        if qasm_output_path is None:
+            self.qasm_output_path = str(resources.files("mqt.benchviewer") / "static/files/qasm_output/")
+        else:
+            self.qasm_output_path = qasm_output_path
 
-    def timeout_handler(_signum, _frame):  # Custom signal handler
-        raise TimeoutException()
+        Path(utils.get_qasm_output_path()).mkdir(exist_ok=True, parents=True)
 
-    # Change the behavior of SIGALRM
-    signal.signal(signal.SIGALRM, timeout_handler)
-    signal.alarm(timeout)
-    try:
-        res = func(*args)
-    except TimeoutException:
-        print(
-            "Calculation/Generation exceeded timeout limit for ",
-            func.__name__,
-            func.__module__.split(".")[-1],
-            args[0].name,
-            args[1:],
-        )
-        return False
-    except Exception as e:
-        print(
-            "Exception: ",
-            e,
-            func.__name__,
-            func.__module__.split(".")[-1],
-            args[0].name,
-            args[1:],
-        )
-        return False
-    else:
-        # Reset the alarm
-        signal.alarm(0)
-
-    return res
+    def set_qasm_output_path(self, new_path: str):
+        self.qasm_output_path = new_path
 
+    def get_qasm_output_path(self):
+        """Returns the path where all .qasm files are stored."""
+        return self.qasm_output_path
 
-def qc_creation_watcher(func, args):
-    class TimeoutException(Exception):  # Custom exception class
-        pass
+    def create_benchmarks_from_config(self) -> bool:
+        Parallel(n_jobs=-1, verbose=100)(
+            delayed(self.generate_benchmark)(benchmark) for benchmark in self.cfg["benchmarks"]
+        )
+        return True
```

Comment:
It's really unfortunate that this function and the `generate_benchmark` function are not covered by tests.
Since they involve so much hard coding it's fairly easy to break these in future updates.
Could you think of some way to run these as tests in CI that doesn't take hours? (Maybe by not trying to generate all levels, but only one of them, e.g., only the independent level)

Reply:
I added a test to cover that method by creating a temporary `config.json` with just one benchmark algorithm with a small size of `2` qubits.